pennylane/transforms/compile.py

```diff
@@ -148,7 +153,8 @@ def qfunc(x, y, z):
             )
         else:
             # Expands out anything that is not a single operation (i.e., the templates)
-            expanded_tape = tape.expand(stop_at=lambda obj: obj.name in all_ops)
+            stop_at = lambda obj: (obj.name in all_ops) and (obj.name != "Barrier")
```

Comment:
As far as I can tell, the point of this line is to expand away Barriers when `only_visual=True`.  Here we differentiate between `only_visual=True` and `only_visual=False` based on whether or not it can be expanded at all.  If `only_visual=False`, then expansion raises a `NotImplementedError`, and so it doesn't get expanded away.

If we explicitly checked `not getattr(obj, "only_visual", False)` for the stopping condition, Barriers with `only_visual=False` will not get expanded at this step, even if they have a defined empty decomposition of `[]`.  

This would also improve readability, especially if we added in a comment as well.  I always find checking for situations explicitly is better than relying on an error getting raised at some later point. More readable and thus more maintainable.

Once we always have decomposition for all scenarios, we would also not have to worry about supporting the operation on all devices and performing extra operations on the state that don't do anything and just waste resources.


```suggestion
            # expand barriers when `only_visual=True`
            stop_at = lambda obj: (obj.name in all_ops) and (not getattr(obj, "only_visual", False))
```